docs: How to setup monitoring stack? (#480)
kinvolk/lokomotive: docs/how-to-guides/monitoring-with-prometheus-operator.md

  * [Install Prometheus Operator](#install-prometheus-operator)
* [Accessing Prometheus, Alertmanager and Grafana](#accessing-prometheus-operator-sub-components)
  * [Accessing Prometheus](#accessing-prometheus)
  * [Accessing Alertmanager](#accessing-alertmanager)
  * [Accessing Grafana](#accessing-grafana)
    * [Using port forward](#using-port-forward)
    * [Using Ingress](#using-ingress)
* [Additional resources](#additional-resources)

## Introduction

This guide provides the steps for deploying a monitoring stack using the `prometheus-operator` Lokomotive component and explains how to access Prometheus, Alertmanager and Grafana.

## Prerequisites

* A Lokomotive cluster deployed on a supported provider and accessible via `kubectl`.

<!---
TODO: Once we have tutorials on how to deploy and configure Rook or OpenEBS, point the following to those tutorials.
-->

* A storage provider component (`rook` and `rook-ceph` or `openebs-operator` and `openebs-storage-class`) deployed with a default storage class that can provision volumes for the [PVCs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) created by Alertmanager and Prometheus.

> > **NOTE**: If you wish to [expose Grafana to the public internet](#using-ingress), the following Lokomotive components should be installed:
> 
> <!---
> TODO: Once we have a tutorial on how to deploy and configure Contour and cert-manager, point the following to that tutorial.
> -->
> 
>   * [`metallb`](./ingress-with-contour-metallb.md) (only on Packet and bare-metal)
>   * [`contour`](../configuration-reference/components/contour.md)
>   * [`cert-manager`](../configuration-reference/components/cert-manager.md)

Reviewer: This looks a bit weird in rendering:
![image](https://user-images.githubusercontent.com/507354/85845904-4bf42600-b7a5-11ea-8cbb-b1cf8d2be56e.png)

Can we make it all in quotation format? I think this should do the trick

```suggestion
<!---
TODO: Once we have a tutorial on how to deploy and configure Contour and cert-manager, point the following to that tutorial.
-->

> **NOTE**: If you wish to [expose Grafana to the public internet](#using-ingress), the following Lokomotive components should be installed:
> * [`metallb`](./ingress-with-contour-metallb.md) (only on Packet and bare-metal)
> * [`contour`](../configuration-reference/components/contour.md)
> * [`cert-manager`](../configuration-reference/components/cert-manager.md)
```

Execute the following command to forward port `9093` locally to the Alertmanager pod:

```bash
kubectl -n monitoring port-forward svc/prometheus-operator-alertmanager 9093
```

Open the following URL: [http://localhost:9093](http://localhost:9093).

### Accessing Grafana

#### Using port forward

Execute the following command to forward port `8080` locally to the Grafana dashboard pod on port `80`:

```bash
kubectl -n monitoring port-forward svc/prometheus-operator-grafana 8080:80
```

Obtain the password for the `admin` Grafana user by running the following command:

```bash
kubectl -n monitoring get secret prometheus-operator-grafana -o jsonpath='{.data.admin-password}' | base64 -d && echo
```

Open the following URL: [http://localhost:8080](http://localhost:8080). Enter the username `admin` and password obtained from the previous step.

#### Using Ingress

To expose Grafana to the internet using Ingress, provide the host field. The configuration for Grafana in the `prometheus-operator` component should look like the following:

```tf
component "prometheus-operator" {
  grafana {
    ingress {
      host = "grafana.<cluster name>.<DNS zone>"
    }
  }
}
```

> > **NOTE**: On Packet, create a DNS entry for `grafana.<cluster name>.<DNS zone>` and point it to the Packet EIP as mentioned [here](./ingress-with-contour-metallb.md).

Reviewer: I don't see anything DNS related here, not sure we have documentation on this yet.

Does it make sense to suggest using External DNS? (we'll have to check it works).

Author: We surely have to work upon our DNS story, at least remove the differences in how we suggest on Packet and on AWS. For AWS things work out of the box.

Reviewer: I just added the ingress block to a prometheus-operator configuration of a cluster with the external DNS component installed, applied, and I could access the grafana dashboard through ingress without any further configuration.

How about we change this to the following?

```suggestion
> **NOTE**: On Packet, you either need to create a DNS entry for `grafana.<cluster name>.<DNS zone>` and point it to the Packet external IP for the contour service (see the [Packet ingress guide for more details](./ingress-with-contour-metallb.md)) or use the [External DNS component](../configuration-reference/components/external-dns.md).
```

Author: done.

Author: done.

# Setting up monitoring on Lokomotive

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Deploy Prometheus Operator](#deploy-prometheus-operator)
  * [Configure Prometheus Operator](#configure-prometheus-operator)
  * [Install Prometheus Operator](#install-prometheus-operator)
* [Accessing Prometheus Operator sub-components](#accessing-prometheus-operator-sub-components)
  * [Accessing Prometheus](#accessing-prometheus)
  * [Accessing Alertmanager](#accessing-alertmanager)
  * [Accessing Grafana](#accessing-grafana)
    * [Using port forward](#using-port-forward)
    * [Using Ingress](#using-ingress)

## Introduction

> This guide provides installation steps of monitoring stack using prometheus-operator component and ways to access the sub-components like Prometheus, Alertmanager and Grafana.

Reviewer 2: We could link here to prometheus-operator component documentation.

Author: Yes it does in the `### Configure Prometheus Operator` section.

Author: Also added it in `Additional resources`.

> This guide provides installation steps of monitoring stack using prometheus-operator component and ways to access the sub-components like Prometheus, Alertmanager and Grafana.

Reviewer 3: ```suggestion
This guide provides the steps for deploying a monitoring stack using the `prometheus-operator` Lokomotive component and explains how to access Prometheus, Alertmanager and Grafana.
```

I suggest this alternative phrasing.

- Add missing articles.
- Avoid using the term "component" twice with separate meanings.
- Use "deployment" instead of "installation".

## Prerequisites

> - A Lokomotive cluster is accessible via `kubectl` deployed on a supported provider.

Reviewer 3: ```suggestion
- A Lokomotive cluster deployed on a supported provider and accessible via `kubectl`.
```

<!---

> TODO: Once we have tutorials on how to deploy and configure rook or openebs, then point the following to those tutorials.

Reviewer 3: ```suggestion
TODO: Once we have tutorials on how to deploy and configure Rook or OpenEBS, point the following to those tutorials.
```

-->

> - Deploy a storage provider component (rook and rook-ceph or openebs-operator and openebs-storage-class) with default storage class that can cater to the [PVC](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) requests made by prometheus-operator components Alertmanager and Prometheus.

Reviewer 2: Again, I'd link to storage components here and maybe provide some sample configuration for them, if we don't have guides to refer to.

Author: I think that's why those todos exist. That this is largely work in progress and we will refer to the guides soon enough.

Author: Also the work to add rook is largely WIP in https://github.com/kinvolk/lokomotive/pull/494 and can be pointed to from here.

> - Deploy a storage provider component (rook and rook-ceph or openebs-operator and openebs-storage-class) with default storage class that can cater to the [PVC](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) requests made by prometheus-operator components Alertmanager and Prometheus.

Reviewer 3: ```suggestion
- A storage provider component (`rook` and `rook-ceph` or `openebs-operator` and `openebs-storage-class`) deployed with a default storage class that can provision volumes for the [PVCs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) created by Alertmanager and Prometheus.
```

- We list prerequisites here so we shouldn't begin sentences with a verb.
- Missing articles.
- The "C" in PVC stands for "claim", which is IMO equivalent to the word "request" here. I'd say that PVCs are *created* rather than *requested*, because the PVC is itself a kind of a request.
- IMO "cater to" doesn't match the tone of a technical document. I've suggested an alternative phrasing.
- I'd backtick any Lokomotive component names to highlight that we're talking about specific components which the user has to deploy.

- [MetalLB](https://metallb.universe.tf/) deployed on the cluster if you wish to [expose Grafana using Ingress](#using-ingress).

  **NOTE**: Required only for the bare metal and Packet providers.

>   Installation instructions for [MetalLB](./ingress-with-contour-metallb.md) component.

Reviewer 3: I find it weird that we break the list to insert the note and the link to the MetalLB doc. I'd rather put the note *after* the list (and mark it with `>` for style).
I'd consider linking to the MetalLB component from the MetalLB bullet itself instead of linking to the MetalLB website.

<!---

> TODO: Once we have a tutorial on how to deploy and configure Contour, then point the following to that tutorial.

Reviewer 3: ```suggestion
TODO: Once we have a tutorial on how to deploy and configure Contour, point the following to that tutorial.
```

-->

> * [Contour](https://projectcontour.io/) deployed on the cluster if you wish to [expose Grafana using Ingress](#using-ingress).

Reviewer 3: Nit: mixed usage of `*` and `-` in the list.

<!---
TODO: Once we have a tutorial on how to deploy and configure cert-manager, then point the following to that tutorial.
-->

> * [cert-manager](https://cert-manager.io/docs/) deployed on the cluster if you wish to [expose Grafana using Ingress](#using-ingress).

Reviewer 3: Maybe we can put one note at the bottom of the prerequisites section which says something like the following:

>NOTE: If you wish to expose Grafana to the public internet, the following Lokomotive components should be installed:
>
>- `metallb`
>- `contour`
>- `cert-manager`

I find this better than putting "conditionals" on most of the items in the prerequisites list.

## Deploy Prometheus Operator

### Configure Prometheus Operator

> Create a file named `monitoring.lokocfg` with the contents below:

Reviewer 3: ```suggestion
Create a file named `monitoring.lokocfg` with the following contents:
```

```tf
component "prometheus-operator" {}
```

> There are many optional parameters one can provide for prometheus-operator to tweak its behaviour. Find information about all the parameters in the prometheus-operator [config reference doc](../configuration-reference/components/prometheus-operator.md).

Reviewer 3: ```suggestion
For information about all the available configuration options for the `prometheus-operator` component, visit the component's [configuration reference](../configuration-reference/components/prometheus-operator.md).
```

### Install Prometheus Operator

> Execute the following command to install prometheus-operator:

Reviewer 3: ```suggestion
Execute the following command to deploy the `prometheus-operator` component:
```

```bash
lokoctl component apply prometheus-operator
```

> Within a few minutes, the prometheus-operator pods will be in `Running` state in `monitoring` namespace. Verify it by executing the following command:

Reviewer 3: ```suggestion
Verify the pods in the `monitoring` namespace are in the `Running` state (this may take a few minutes):
```

```bash
kubectl -n monitoring get pods
```

> ## Accessing Prometheus Operator sub-components

Reviewer 3: ```suggestion
## Accessing Prometheus, Alertmanager and Grafana
```

### Accessing Prometheus

Execute the following command to forward port `9090` locally to the Prometheus pod:

```bash
kubectl -n monitoring port-forward svc/prometheus-operator-prometheus 9090
```

Open the following URL: [http://localhost:9090](http://localhost:9090).

### Accessing Alertmanager

Execute the following command to forward port `9093` locally to the Alertmanager pod:

```bash
kubectl -n monitoring port-forward svc/prometheus-operator-alertmanager 9093
```

Open the following URL: [http://localhost:9093](http://localhost:9093).

### Accessing Grafana

#### Using port forward

Execute the following command to forward port `8080` locally to the Grafana dashboard pod on port `80`:

```bash
kubectl -n monitoring port-forward svc/prometheus-operator-grafana 8080:80
```

Obtain the password for the `admin` Grafana user by running the following command:

```bash
kubectl -n monitoring get secret prometheus-operator-grafana -o jsonpath='{.data.admin-password}' | base64 -d && echo
```

Open the following URL: [http://localhost:8080](http://localhost:8080). Enter the username `admin` and password obtained from the previous step.

#### Using Ingress

> To expose grafana to the internet using Ingress, provide the host field. The configuration for grafana in prometheus-operator should look like the following:

Reviewer 3: Could you fix the caps please?

```tf
component "prometheus-operator" {
  grafana {
    ingress {
      host = "grafana.<cluster name>.<DNS zone>"
    }
  }
}
```

> **NOTE**: On Packet make a DNS entry for `grafana.<cluster name>.<DNS zone>` against the Packet EIP as mentioned in the
> ["Setting up an HTTP ingress controller on Lokomotive with MetalLB and Contour on Packet"](./ingress-with-contour-metallb.md) guide.

Reviewer 3: ```suggestion
>NOTE: On Packet, create a DNS entry for `grafana.<cluster name>.<DNS zone>` and point it to the Packet EIP as mentioned [here](./ingress-with-contour-metallb.md).
```

Reviewer 3: Again, I feel like we should fix the issue which makes Packet a "special case". I realize this is out of scope though :shrug:

Obtain the password for the `admin` Grafana user by running the following command:

```bash
kubectl -n monitoring get secret prometheus-operator-grafana -o jsonpath='{.data.admin-password}' | base64 -d && echo
```

> Open the following URL: `https://grafana.<cluster name>.<DNS zone>`. Enter the username `admin` and password obtained from the previous step.

Reviewer 3: Could you please add the missing article?
